src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -0,0 +1,823 @@
+                       }];
+}
+
+// TODO: support more than one subscriber / subscription delegate
+//   * reconcile second+ subscriptions with different min and max intervals
+//     (save min/max intervals for each subscriber and resubscribe when they change)
+//   * events should be cached and replayed for second+ subscribers
+- (void)subscribeWithDelegate:(id<MTRDeviceSubscriptionDelegate>)delegate
+                        queue:(dispatch_queue_t)queue
+                  minInterval:(uint16_t)minInterval
+                  maxInterval:(uint16_t)maxInterval
+                       params:(MTRSubscribeParams *)params
+{
+    // Save a weak reference to delegate, so if/when the object goes away, they are automatically unsubscribed
+    MTRWeakReference<id<MTRDeviceSubscriptionDelegate>> * weakDelegate = [MTRWeakReference weakReferenceWithObject:delegate];
+    [_subscribers addObject:weakDelegate];
+
+    // perform the actual subscription
+    [self subscribeWithMinInterval:minInterval maxInterval:maxInterval params:params];
+}
+
+#pragma mark Device Interactions
+- (NSDictionary<NSString *, id> *)readAttributeWithEndpointId:(NSNumber * _Nullable)endpointId
+                                                    clusterId:(NSNumber * _Nullable)clusterId
+                                                  attributeId:(NSNumber * _Nullable)attributeId
+                                                       params:(MTRReadParams * _Nullable)params
+{
+    // Create work item, set ready handler to perform task, then enqueue the work
+    MTRAsyncCallbackQueueWorkItem * workItem = [[MTRAsyncCallbackQueueWorkItem alloc] initWithQueue:_queue];
+    MTRAsyncCallbackReadyHandler readyHandler = ^(MTRDevice * device, NSUInteger retryCount) {
+        [self connectAndPerformAsync:^(MTRBaseDevice * baseDevice) {
+            [baseDevice readAttributeWithEndpointId:endpointId
+                                          clusterId:clusterId
+                                        attributeId:attributeId
+                                             params:params
+                                        clientQueue:self.queue
+                                         completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values,
+                                             NSError * _Nullable error) {
+                                             if (values) {
+                                                 // Since the format is the same data-value dictionary, this looks like an attribute
+                                                 // report
+                                                 [self _handleAttributeReport:values];
+                                             }
+                                             // TODO: retry on error
+                                             [workItem endWork];
+                                         }];
+        }];
+    };
+    workItem.readyHandler = readyHandler;
```

Comment:
My knowledge of ARC is pitiful, so maybe this entire comment is wrong....

At this point `workItem` is holding a strong ref to `readyHandler`.   `readyHandler` closes over `workItem`, so is holding a strong ref to `workItem`.  Where is this cycle broken so we don't leak?